### ISSUE TITLE
Issue with OSX

### ISSUE THREAD

**Issue by SenWu** [HazyResearch/metal#28]
Hi,

I got this issue when running MeTaL on Mac OS X.
```
E   RuntimeError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are using (Ana)Conda please install python.app and replace the use of 'python' with 'pythonw'. See 'Working with Matplotlib on OSX' in the Matplotlib FAQ for more information.
```

There are some fixes online using either different version of python or modified python, but they are not ideal case. Is there a proper way to get rid of this issue?

**Comment by SenWu** [comment]
This seems useful: https://stackoverflow.com/questions/34977388/matplotlib-runtimeerror-python-is-not-installed-as-a-framework

**Comment by bhancock8** [comment]
Fixed by #32.

**Comment by bhancock8** [comment]
Hmm, on second look it looks like this dragon is not so easily slain. We now satisfy pip (i.e. `pip install snorkel-metal` works correctly), but if somebody sets `%matplotlib inline` in a jupyter notebook before they've imported mtal, then we currently get a nasty warning that the `matplotlib.use("TkAgg")` command I added to the repo does nothing. Nothing breaks, but it ain't pretty. I'll look for another solution.

**Comment by lukehsiao** [comment]
I also get the above error when loading metal in a notebook.

```
home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/metal/analysis.py:13: UserWarning: 
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

The backend was *originally* set to 'module://ipykernel.pylab.backend_inline' by the following code:
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/kernelapp.py", line 486, in start
    self.io_loop.start()
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/tornado/platform/asyncio.py", line 132, in start
    self.asyncio_loop.run_forever()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 422, in run_forever
    self._run_once()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 1432, in _run_once
    handle._run()
  File "/usr/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/tornado/platform/asyncio.py", line 122, in _handle_events
    handler_func(fileobj, events)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py", line 450, in _handle_events
    self._handle_recv()
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py", line 480, in _handle_recv
    self._run_callback(callback, msg)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py", line 432, in _run_callback
    callback(*args, **kwargs)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 283, in dispatcher
    return self.dispatch_shell(stream, msg)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 233, in dispatch_shell
    handler(stream, idents, msg)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 399, in execute_request
    user_expressions, allow_stdin)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/ipkernel.py", line 208, in do_execute
    res = shell.run_cell(code, store_history=store_history, silent=silent)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/zmqshell.py", line 537, in run_cell
    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2666, in run_cell
    self.events.trigger('post_run_cell', result)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/IPython/core/events.py", line 88, in trigger
    func(*args, **kwargs)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/ipykernel/pylab/backend_inline.py", line 160, in configure_once
    activate_matplotlib(backend)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/IPython/core/pylabtools.py", line 311, in activate_matplotlib
    matplotlib.pyplot.switch_backend(backend)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/matplotlib/pyplot.py", line 231, in switch_backend
    matplotlib.use(newbackend, warn=False, force=True)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/matplotlib/__init__.py", line 1422, in use
    reload(sys.modules['matplotlib.backends'])
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/importlib/__init__.py", line 166, in reload
    _bootstrap._exec(spec, module)
  File "/home/lwhsiao/repos/tutorials/.venv/lib/python3.6/site-packages/matplotlib/backends/__init__.py", line 16, in <module>
    line for line in traceback.format_stack()


  matplotlib.use("TkAgg")
```

**Comment by bhancock8** [comment]
Ok, this clearly needs a different solution then, 'cause that warning, while not breaking, is long and obnoxious. We need something that lets metal use matplotlib, build successfully on pip with it in the requirements, and not output this warning when somebody imports matplotlib before importing metal. I'm traveling with limited internet, and won't be able to get to this until probably late next week at the earliest. If anyone else has a sec to take a shot at it, a PR is welcome!

**Comment by bhancock8** [comment]
Apparently matplotlib is still causing import issues. Looking at other libraries and threads, I saw this comment on scikit-optimize (https://github.com/scikit-optimize/scikit-optimize/issues/637) and I agree:

> There are good reasons for why scientific libraries don't usually have matplotlib as a direct dependency: it is very platform dependent and it is usually not essential for what these libraries are trying to accomplish. Why should an optimization library depend so heavily on a visualization library?
> 
> I agree that you shouldn't try to fix dependencies for the user and I would go even further: you shouldn't even install it for them.
> 
> Take pandas as example: they have the plot() method on dataframes and series that call matplotlib.pyplot.plot() but they only import it at the time of the call and matplotlib isn't even a dependency of pandas! And the reason is very simple: pandas is a data manipulation library, the visualization part is just a kind of add-on. Leaving matplotlib out by default instead of pushing it into users that don't even need it seems like a reasonable decision to make.
> 
> I'm trying to solve a similar problem to the one described. We have a library that will depend on scikit-optimize but we are having problems because we can't depend on matplotlib platform sensitiveness and, sincerely, we just need the optimizing functions.

So I propose that we make a contrib/visualization directory. Any helpers we have that use matplotlib (all but one are in analysis.py, and that last is three-liner method we don't currently use anywhere) can be moved there and we'll add matplotlib to that directory's extra requirements.txt and remove it from the core/pip dependencies. We'll keep a notebook showing them but won't make it part of the core tests that run with travis. It's a heavy and tricky enough library and peripheral enough to metal's core functionality that it'll be nice to have proper setup for it off our plate. @ajratner agree?

**Comment by ajratner** [comment]
@bhancock8 I like this!  Also (because somewhat related): I had actually been thinking that showing the visualizations in the Basics tutorial, while nice, might give the (misleading) impression that lots of manual visualization is needed as a first step to using our approach.  (Granted, it often is needed in iteration- but the idea is that it absolutely shouldn't be on first pass!).  So I'm all for this, for multiple reasons!

**Comment by bhancock8** [comment]
Fixed by #62
